src/main/java/net/fabricmc/loom/util/MojmapCreator.java

```diff
@@ -0,0 +1,244 @@
+import org.cadixdev.lorenz.model.ClassMapping;
+import org.cadixdev.lorenz.model.FieldMapping;
+import org.cadixdev.lorenz.model.InnerClassMapping;
+import org.cadixdev.lorenz.model.MethodMapping;
+import org.cadixdev.lorenz.model.MethodParameterMapping;
+import org.cadixdev.lorenz.model.TopLevelClassMapping;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.internal.artifacts.DefaultModuleIdentifier;
+import org.gradle.api.internal.artifacts.dependencies.DefaultSelfResolvingDependency;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.internal.component.external.model.DefaultModuleComponentIdentifier;
+import org.zeroturnaround.zip.ByteSource;
+import org.zeroturnaround.zip.ZipEntrySource;
+import org.zeroturnaround.zip.ZipUtil;
+
+import net.fabricmc.loom.LoomGradleExtension;
+import net.fabricmc.loom.providers.MinecraftProvider;
+import net.fabricmc.lorenztiny.TinyMappingsReader;
+import net.fabricmc.mapping.tree.TinyMappingFactory;
+
+public class MojmapCreator {
+	private static final Gson GSON = new Gson();
+
+	public static Dependency mojmap(Project project, LoomGradleExtension extension, String minecraftVersionRaw) throws IOException {
+		File userCache = extension.getUserCache();
+		File mappingsDir = new File(userCache, "mappings");
+		File steps = new File(mappingsDir, "steps");
+
+		ManifestVersion.Versions version = MinecraftProvider.findVersion(project, extension, minecraftVersionRaw).get();
+		String id = version.id;
+
+		File mappingsJar = new File(mappingsDir, "mojmap-" + id + ".jar");
+
+		if (!mappingsJar.exists() || project.getGradle().getStartParameter().isRefreshDependencies()) {
+			MappingSet officialToNamed;
+			MappingSet intermediaryToOfficial;
+
+			// Download and parse Proguard logs
+			{
+				File minecraftJson = new File(userCache, "minecraft-" + id + "-info.json");
+				DownloadUtil.downloadIfChanged(new URL(version.url), minecraftJson, project.getLogger());
+
+				MinecraftVersionInfo info = GSON.fromJson(MoreFiles.asCharSource(minecraftJson.toPath(), StandardCharsets.UTF_8).read(), MinecraftVersionInfo.class);
+
+				File clientMappings = new File(steps, "mojmap-" + id + "-client.txt");
+				File serverMappings = new File(steps, "mojmap-" + id + "-server.txt");
+				String clientMappingsUrl = info.downloads.get("client_mappings").url;
+				String serverMappingsUrl = info.downloads.get("server_mappings").url;
+
+				DownloadUtil.downloadIfChanged(new URL(clientMappingsUrl), clientMappings, project.getLogger());
+				DownloadUtil.downloadIfChanged(new URL(serverMappingsUrl), serverMappings, project.getLogger());
+
+				MappingSet mappings = MappingSet.create();
+
+				// Print license headers
+				String clientMappingsString = MoreFiles.asCharSource(clientMappings.toPath(), StandardCharsets.UTF_8).read();
+				String serverMappingsString = MoreFiles.asCharSource(serverMappings.toPath(), StandardCharsets.UTF_8).read();
```

Comment:
Use buffered readers, I think you can mark and then reset back to the beggining of the file after the license has been printed.

Reply:
Marking with a large limit would be equivalent to just reading it to a buffer (this case a string) anyway, I don't see any benefits to this